app/objects/secondclass/c_link.py

```diff
@@ -226,31 +226,49 @@ async def _load_module(module_type, module_info):
-
-    @staticmethod
-    def _fact_exists(new_fact, fact):
-        return new_fact.trait == fact.trait and new_fact.value == fact.value
-
-    def _is_new_host_fact(self, new_fact, fact):
-        return new_fact.trait[:5] == 'host.' and self.paw != fact.collected_by
-
-    async def _update_scores(self, operation, increment):
-        for uf in self.used:
-            all_facts = operation.all_facts() if operation else self.facts
-            for found_fact in all_facts:
-                if found_fact.unique == uf.unique:
-                    found_fact.score += increment
-                    break
+    async def _save_fact(self, operation, fact, score, relationship):
+        knowledge_svc_handle = BaseService.get_service('knowledge_svc')
+        all_facts = await operation.all_facts() if operation else self.facts
+        source = operation.id if operation else self.id
+        rl = [relationship] if relationship else []
+        if all([fact.trait, fact.value]):
+            if not await knowledge_svc_handle.check_fact_exists(fact, all_facts):
+                f_gen = Fact(trait=fact.trait, value=fact.value, source=source, score=score, collected_by=self.paw,
+                             technique_id=self.ability.technique_id, links=[self.id], relationships=rl,
+                             origin_type=OriginType.LEARNED)
+                self.facts.append(f_gen)
+                await knowledge_svc_handle.add_fact(f_gen)
+            else:
+                existing_fact = (await knowledge_svc_handle.get_facts(criteria=dict(trait=fact.trait,
+                                                                                    value=fact.value)))[0]
+                existing_fact.links.append(self.id)
+                if relationship not in existing_fact.relationships:
+                    existing_fact.relationships.append(relationship)
+                await knowledge_svc_handle.update_fact(criteria=dict(trait=fact.trait, value=fact.value),
+                                                       updates=dict(links=existing_fact.links,
+                                                                    relationships=existing_fact.relationships))
+                existing_local_record = [x for x in self.facts if x.trait == fact.trait and x.value == fact.value]
+                if existing_local_record:
+                    existing_local_record[0].links = existing_fact.links
+                else:
+                    self.facts.append(existing_fact)
+
+
+async def _update_scores(operation, increment, used, facts):
+    knowledge_svc_handle = BaseService.get_service('knowledge_svc')
+    for uf in used:
+        all_facts = await operation.all_facts() if operation else facts
+        for found_fact in all_facts:
+            if found_fact.unique == uf.unique:
+                found_fact.score += increment
+                await knowledge_svc_handle.update_fact(dict(trait=found_fact.trait, value=found_fact.value,
```

Comment:
Do we need to update the fact here, or is the found_fact object the same object we are updating in the knowledge store?  (though I may answer my own question and say it is good to do since if we swap out the underlaying data storage we cant rely on the fact python is passing around the mutable objects)

Reply:
Yep, that's pretty much it. I'm pretty sure that if we're using the base python implementation of the backend, we could just update the object in place, but if we're going to be compatible with any of the other possible backends, we can't make that assumption because the datastore methods may vary significantly at that point.